File: backend/src/main/java/com/school/management/entity/task/Task.java
package com.school.management.entity.task;

import com.baomidou.mybatisplus.annotation.TableField;
import com.baomidou.mybatisplus.annotation.TableName;
import com.baomidou.mybatisplus.annotation.Version;
import com.baomidou.mybatisplus.extension.handlers.JacksonTypeHandler;
import com.school.management.entity.BaseEntity;
import lombok.Data;
import lombok.EqualsAndHashCode;

import java.time.LocalDateTime;
import java.util.List;

/**
 * 任务实体
 */
@Data
@EqualsAndHashCode(callSuper = true)
@TableName(value = "tasks", autoResultMap = true)
public class Task extends BaseEntity {

    /**
     * 任务编号(如: TASK-20251227-0001)
     */
    private String taskCode;

    /**
     * 任务标题
     */
    private String title;

    /**
     * 任务描述
     */
    private String description;

    /**
     * 优先级: 1-紧急, 2-普通, 3-低
     */
    private Integer priority;

    /**
     * 分配人ID(创建任务的领导)
     */
    private Long assignerId;

    /**
     * 分配人姓名
     */
    private String assignerName;

    /**
     * 分配类型: 1-指定个人, 2-批量分配
     */
    private Integer assignType;

    /**
     * 部门ID(任务所属部门)
     */
    private Long departmentId;

    /**
     * 截止时间
     */
    private LocalDateTime dueDate;

    /**
     * 任务状态: 0-待接收, 1-进行中, 2-已提交, 3-已完成, 4-已拒绝
     */
    private Integer status;

    /**
     * 目标ID列表(批量分配时的执行人ID列表)
     */
    @TableField(typeHandler = JacksonTypeHandler.class)
    private List<Long> targetIds;

    /**
     * 执行人ID(单人任务时)
     */
    @TableField(exist = false)
    private Long assigneeId;

    /**
     * 接受时间
     */
    private LocalDateTime acceptedAt;

    /**
     * 提交时间
     */
    private LocalDateTime submittedAt;

    /**
     * 完成时间
     */
    private LocalDateTime completedAt;

    /**
     * 流程模板ID
     */
    private Long workflowTemplateId;

    /**
     * 流程实例ID
     */
    private String processInstanceId;

    /**
     * 当前审批节点
     */
    private String currentNode;

    /**
     * 当前审批人列表
     */
    @TableField(typeHandler = JacksonTypeHandler.class)
    private List<Long> currentApprovers;

    /**
     * 任务附件ID列表
     */
    @TableField(typeHandler = JacksonTypeHandler.class)
    private List<Long> attachmentIds;

    /**
     * 乐观锁版本号
     * 用于防止并发修改冲突，每次更新时自动递增
     */
    @Version
    private Integer version;

    // ========== 非数据库字段 ==========

    /**
     * 执行人列表(批量分配时)
     */
    @TableField(exist = false)
    private List<TaskAssignee> assignees;

    /**
     * 审批记录列表
     */
    @TableField(exist = false)
    private List<TaskApprovalRecord> approvalRecords;
}

File: frontend/src/types/v2/task.ts
/**
 * 任务管理类型定义 - DDD架构适配
 */

// ==================== 枚举类型 ====================

/**
 * 任务状态
 */
export type TaskStatus = 0 | 1 | 2 | 3 | 4 | 5 | 6

export const TaskStatusMap: Record<TaskStatus, string> = {
  0: '待接收',
  1: '进行中',
  2: '待审核',
  3: '已完成',
  4: '已打回',
  5: '已取消',
  6: '已超期'
}

/**
 * 任务优先级
 */
export type TaskPriority = 1 | 2 | 3

export const TaskPriorityMap: Record<TaskPriority, string> = {
  1: '紧急',
  2: '普通',
  3: '低'
}

/**
 * 分配类型
 */
export type AssignType = 1 | 2

export const AssignTypeMap: Record<AssignType, string> = {
  1: '指定个人',
  2: '批量分配'
}

/**
 * 审批动作
 */
export type ApprovalAction = 1 | 2

export const ApprovalActionMap: Record<ApprovalAction, string> = {
  1: '通过',
  2: '打回'
}

// ==================== 实体类型 ====================

/**
 * 任务提交记录
 */
export interface TaskSubmission {
  id: number
  taskId: number
  assigneeId: number
  content: string
  attachmentIds?: number[]
  submittedAt: string
  createdAt?: string
}

/**
 * 任务审批记录
 */
export interface TaskApprovalRecord {
  id: number
  taskId: number
  submissionId: number
  approverId: number
  approverName: string
  approvalLevel: number
  action: ApprovalAction
  actionText?: string
  comment?: string
  createdAt: string
}

/**
 * 任务执行人
 */
export interface TaskAssignee {
  id: number
  taskId: number
  assigneeId: number
  assigneeName: string
  orgUnitId?: number
  orgUnitName?: string
  status: TaskStatus
  statusText?: string
  acceptedAt?: string
  submittedAt?: string
  completedAt?: string
  processInstanceId?: string
  currentApprovalLevel?: number
  submission?: TaskSubmission
  approvalRecords?: TaskApprovalRecord[]
}

/**
 * 任务进度节点
 */
export interface TaskProgressNode {
  nodeId: string
  nodeName: string
  nodeType: string
  status: 'pending' | 'active' | 'completed' | 'rejected'
  operatorId?: number
  operatorName?: string
  operatedAt?: string
  comment?: string
}

/**
 * 任务实体
 */
export interface Task {
  id: number
  taskCode: string
  title: string
  description?: string
  priority: TaskPriority
  priorityText?: string
  status: TaskStatus
  statusText?: string
  assignerId: number
  assignerName: string
  assignType: AssignType
  assigneeId?: number
  assigneeName?: string
  orgUnitId?: number
  orgUnitName?: string
  assignees?: TaskAssignee[]
  totalAssignees?: number
  submittedAssignees?: number
  completedAssignees?: number
  myAssigneeId?: number
  myStatus?: TaskStatus
  myAcceptedAt?: string
  mySubmittedAt?: string
  dueDate?: string
  overdue?: boolean
  acceptedAt?: string
  submittedAt?: string
  completedAt?: string
  workflowTemplateId?: number
  workflowTemplateName?: string
  processInstanceId?: string
  currentNode?: string
  currentApprovers?: number[]
  attachmentIds?: number[]
  submission?: TaskSubmission
  approvalRecords?: TaskApprovalRecord[]
  createdAt: string
  updatedAt?: string
}

/**
 * 任务详情（含卡片式执行人数据）
 */
export interface TaskDetail extends Task {
  assigneeCards?: TaskAssigneeCard[]
}

/**
 * 执行人卡片数据
 */
export interface TaskAssigneeCard {
  assignee: TaskAssignee
  submission?: TaskSubmission
  approvalRecords?: TaskApprovalRecord[]
  progressNodes?: TaskProgressNode[]
}

/**
 * 任务统计
 */
export interface TaskStatistics {
  totalCount: number
  pendingCount: number
  inProgressCount: number
  submittedCount: number
  completedCount: number
  rejectedCount: number
  cancelledCount: number
  overdueCount: number
  completionRate: number
  overdueRate: number
  pendingApprovalCount: number
}

/**
 * 工作流模板
 */
export interface WorkflowTemplate {
  id: number
  name: string
  description?: string
  approvalLevels: number
  levelConfigs?: ApprovalLevelConfig[]
  status: number
  createdAt?: string
  updatedAt?: string
}

/**
 * 审批级别配置
 */
export interface ApprovalLevelConfig {
  level: number
  name: string
  approverType: 'role' | 'user' | 'orgUnit'
  approverIds?: number[]
}

/**
 * 组织单元审批配置
 */
export interface OrgUnitApprovalConfig {
  orgUnitId: number
  orgUnitName?: string
  approverConfigs: ApproverConfig[]
}

/**
 * 审批人配置
 */
export interface ApproverConfig {
  level: number
  approverId: number
  approverName?: string
}

// ==================== 请求类型 ====================

/**
 * 创建任务请求
 */
export interface CreateTaskRequest {
  title: string
  description?: string
  priority?: TaskPriority
  assignType: AssignType
  assigneeId?: number
  targetIds?: number[]
  orgUnitId?: number
  dueDate?: string
  workflowTemplateId: number
  approvalConfigs: OrgUnitApprovalConfig[]
  attachmentIds?: number[]
}

/**
 * 任务查询参数
 */
export interface TaskQueryParams {
  keyword?: string
  status?: TaskStatus
  priority?: TaskPriority
  assignerId?: number
  assigneeId?: number
  orgUnitId?: number
  startTime?: string
  endTime?: string
  dueStartTime?: string
  dueEndTime?: string
  myTask?: boolean
  pendingApproval?: boolean
  pageNum?: number
  pageSize?: number
}

/**
 * 任务提交请求
 */
export interface TaskSubmitRequest {
  taskId: number
  content: string
  attachmentIds?: number[]
}

/**
 * 任务审批请求
 */
export interface TaskApproveRequest {
  taskId: number
  submissionId: number
  action: ApprovalAction
  comment?: string
  rejectToNode?: string
  flowableTaskId?: string
}

/**
 * 创建工作流模板请求
 */
export interface CreateWorkflowTemplateRequest {
  name: string
  description?: string
  approvalLevels: number
  levelConfigs?: ApprovalLevelConfig[]
}

/**
 * 更新工作流模板请求
 */
export interface UpdateWorkflowTemplateRequest extends Partial<CreateWorkflowTemplateRequest> {
  id?: number
}

// ==================== 系统消息 ====================

/**
 * 系统消息
 */
export interface SystemMessage {
  id: number
  userId: number
  type: string
  title: string
  content: string
  relatedId?: number
  relatedType?: string
  isRead: boolean
  createdAt: string
}

/**
 * 系统消息查询参数
 */
export interface SystemMessageQueryParams {
  type?: string
  isRead?: boolean
  pageNum?: number
  pageSize?: number
}
